### ISSUE TITLE
“NoSuchMethodErrors” due to multiple versions of com.google.guava:guava:jar

### ISSUE THREAD

**Issue by Bing-ok** [wso2/micro-integrator#1953]
## Issue description
Hi, there are multiple versions of **_com.google.guava:guava_** in **_micro-integrator\components\org.wso2.micro.integrator.server_**. As shown in the following dependency tree, library **_com.google.guava:guava:22.0_** is transitively introduced by **_org.wso2.config.mapper:config-mapper:1.0.12_**, but has been managed to be version **_27.0-jre_**.

However, several methods defined in shadowed version **_com.google.guava:guava:22.0_** are referenced by client project via **_org.wso2.config.mapper:config-mapper:1.0.12_**, but missing in the actually loaded version **_com.google.guava:guava:27.0-jre_**.

For instance, the following missing methods(defined in **_com.google.guava:guava:22.0_**) are actually referenced by **_micro-integrator\components\org.wso2.micro.integrator.server_**, which will introduce a runtime error(i.e., "NoSuchMethodErrors") into **_micro-integrator\components\org.wso2.micro.integrator.server_**.
**_1. com.google.common.collect.Streams: java.util.stream.LongStream lambda$concat$2(java.util.stream.LongStream)_** is invoked by **_micro-integrator\components\org.wso2.micro.integrator.server_** via the following path:
<pre><code>
paths------
&lt;org.wso2.micro.integrator.server.Main: void main(java.lang.String[])&gt; micro-integrator\components\org.wso2.micro.integrator.server\target\classes
&lt;org.wso2.micro.integrator.server.Main: void handleConfiguration()&gt; micro-integrator\components\org.wso2.micro.integrator.server\target\classes
&lt;org.wso2.config.mapper.ConfigParser: void parse(java.lang.String,java.lang.String,java.lang.String)&gt; Repositories\org\wso2\config\mapper\config-mapper\1.0.12\config-mapper-1.0.12.jar
&lt;com.google.common.collect.LinkedListMultimap$1EntriesImpl: void forEach(java.util.function.Consumer)&gt; 
&lt;com.google.common.collect.CollectSpliterators$1$lambda_forEachRemaining_1__42: void accept(java.lang.Object)&gt; 
&lt;com.google.common.collect.CollectSpliterators$1: void lambda$forEachRemaining$1(java.util.function.Consumer,java.util.function.Function,java.lang.Object)&gt; 
&lt;com.google.common.collect.Streams$lambda_concat_2__169: java.lang.Object apply(java.lang.Object)&gt; 
&lt;com.google.common.collect.Streams: java.util.stream.LongStream lambda$concat$2(java.util.stream.LongStream)&gt;
</code></pre>

**_2. com.google.common.collect.Streams: java.util.stream.IntStream lambda$concat$1(java.util.stream.IntStream)_** is invoked by **_micro-integrator\components\org.wso2.micro.integrator.server_** via the following path:
<pre><code>
paths------
&lt;org.wso2.micro.integrator.server.Main: void main(java.lang.String[])&gt; micro-integrator\components\org.wso2.micro.integrator.server\target\classes
&lt;org.wso2.micro.integrator.server.Main: void handleConfiguration()&gt; micro-integrator\components\org.wso2.micro.integrator.server\target\classes
&lt;org.wso2.config.mapper.ConfigParser: void parse(java.lang.String,java.lang.String,java.lang.String)&gt; Repositories\org\wso2\config\mapper\config-mapper\1.0.12\config-mapper-1.0.12.jar
&lt;com.google.common.collect.LinkedListMultimap$1EntriesImpl: void forEach(java.util.function.Consumer)&gt; 
&lt;com.google.common.collect.CollectSpliterators$1$lambda_forEachRemaining_1__42: void accept(java.lang.Object)&gt; 
&lt;com.google.common.collect.CollectSpliterators$1: void lambda$forEachRemaining$1(java.util.function.Consumer,java.util.function.Function,java.lang.Object)&gt; 
&lt;com.google.common.collect.Streams$lambda_concat_1__168: java.lang.Object apply(java.lang.Object)&gt;  
&lt;com.google.common.collect.Streams: java.util.stream.IntStream lambda$concat$1(java.util.stream.IntStream)&gt;
</code></pre>

**_3. com.google.common.collect.Streams: void lambda$forEachPair$4(java.util.function.BiConsumer,com.google.common.collect.Streams$TemporaryPair)_** is invoked by **_micro-integrator\components\org.wso2.micro.integrator.server_** via the following path:
<pre><code>
paths------
&lt;org.wso2.micro.integrator.server.Main: void main(java.lang.String[])&gt; micro-integrator\components\org.wso2.micro.integrator.server\target\classes
&lt;org.wso2.micro.integrator.server.Main: void handleConfiguration()&gt; micro-integrator\components\org.wso2.micro.integrator.server\target\classes
&lt;org.wso2.config.mapper.ConfigParser: void parse(java.lang.String,java.lang.String,java.lang.String)&gt; Repositories\org\wso2\config\mapper\config-mapper\1.0.12\config-mapper-1.0.12.jar
&lt;com.google.common.collect.LinkedListMultimap$1EntriesImpl: void forEach(java.util.function.Consumer)&gt; 
&lt;com.google.common.collect.Streams$lambda_forEachPair_4__172: void accept(java.lang.Object)&gt; 
&lt;com.google.common.collect.Streams: void lambda$forEachPair$4(java.util.function.BiConsumer,com.google.common.collect.Streams$TemporaryPair)&gt;
</code></pre>

**_4. com.google.common.collect.Streams: java.util.stream.DoubleStream lambda$concat$3(java.util.stream.DoubleStream)_** is invoked by **_micro-integrator\components\org.wso2.micro.integrator.server_** via the following path:
<pre><code>
paths------
&lt;org.wso2.micro.integrator.server.Main: void main(java.lang.String[])&gt; micro-integrator\components\org.wso2.micro.integrator.server\target\classes
&lt;org.wso2.micro.integrator.server.Main: void handleConfiguration()&gt; micro-integrator\components\org.wso2.micro.integrator.server\target\classes
&lt;org.wso2.config.mapper.ConfigParser: void parse(java.lang.String,java.lang.String,java.lang.String)&gt; Repositories\org\wso2\config\mapper\config-mapper\1.0.12\config-mapper-1.0.12.jar
&lt;com.google.common.collect.LinkedListMultimap$1EntriesImpl: void forEach(java.util.function.Consumer)&gt; 
&lt;com.google.common.collect.CollectSpliterators$1$lambda_forEachRemaining_1__42: void accept(java.lang.Object)&gt; 
&lt;com.google.common.collect.CollectSpliterators$1: void lambda$forEachRemaining$1(java.util.function.Consumer,java.util.function.Function,java.lang.Object)&gt; 
&lt;com.google.common.collect.Streams$lambda_concat_3__170: java.lang.Object apply(java.lang.Object)&gt;  
&lt;com.google.common.collect.Streams: java.util.stream.DoubleStream lambda$concat$3(java.util.stream.DoubleStream)&gt;
</code></pre>

**_5. com.google.common.collect.Iterables$5: void lambda$forEach$0(java.lang.Class,java.util.function.Consumer,java.lang.Object)_** is invoked by **_micro-integrator\components\org.wso2.micro.integrator.server_** via the following path
<pre><code>
paths------
&lt;org.wso2.micro.integrator.server.Main: void main(java.lang.String[])&gt; micro-integrator\components\org.wso2.micro.integrator.server\target\classes
&lt;org.wso2.micro.integrator.server.Main: void handleConfiguration()&gt; micro-integrator\components\org.wso2.micro.integrator.server\target\classes
&lt;org.wso2.config.mapper.ConfigParser: void parse(java.lang.String,java.lang.String,java.lang.String)&gt; Repositories\org\wso2\config\mapper\config-mapper\1.0.12\config-mapper-1.0.12.jar
&lt;com.google.common.collect.LinkedListMultimap$1EntriesImpl: void forEach(java.util.function.Consumer)&gt; 
&lt;com.google.common.collect.Iterables$5$lambda_forEach_0__114: void accept(java.lang.Object)&gt; 
&lt;com.google.common.collect.Iterables$5: void lambda$forEach$0(java.lang.Class,java.util.function.Consumer,java.lang.Object)&gt;
</code></pre>

**_6. com.google.common.collect.Iterables$6: void lambda$forEach$0(java.util.function.Consumer,com.google.common.base.Function,java.lang.Object)_** is invoked by **_micro-integrator\components\org.wso2.micro.integrator.server_** via the following path:
<pre><code>
paths------
&lt;org.wso2.micro.integrator.server.Main: void main(java.lang.String[])>&gt; micro-integrator\components\org.wso2.micro.integrator.server\target\classes
&lt;org.wso2.micro.integrator.server.Main: void handleConfiguration()&gt; micro-integrator\components\org.wso2.micro.integrator.server\target\classes
&lt;org.wso2.config.mapper.ConfigParser: void parse(java.lang.String,java.lang.String,java.lang.String)&gt; Repositories\org\wso2\config\mapper\config-mapper\1.0.12\config-mapper-1.0.12.jar
&lt;com.google.common.collect.LinkedListMultimap$1EntriesImpl: void forEach(java.util.function.Consumer)&gt; 
&lt;com.google.common.collect.Iterables$6$lambda_forEach_0__116: void accept(java.lang.Object)&gt; 
&lt;com.google.common.collect.Iterables$6: void lambda$forEach$0(java.util.function.Consumer,com.google.common.base.Function,java.lang.Object)&gt;
</code></pre>

## Suggested fixing solutions
1. Upgrade dependency **_com.hubspot.jinjava:jinjava_** from **_2.4.14_** to **_2.4.15_**. Because the newer version **_com.hubspot.jinjava:jinjava:2.4.15_** does not invoke the above missing methods, such upgrading can solve the problem.
2. Declare a direct dependency **_com.google.guava:guava:22.0_** in the pom file of **_micro-integrator\components\org.wso2.micro.integrator.server_**, to override **_com.google.guava:guava_**'s managed version.

Thank you very much for your attention.
Best regards,

## Dependency tree----
<pre><code>
[INFO] org.wso2.ei:org.wso2.micro.integrator.server:jar:1.3.0-SNAPSHOT
[INFO] +- org.ops4j.pax.logging:pax-logging-api:jar:1.11.2:compile
[INFO] |  +- org.osgi:osgi.core:jar:6.0.0:compile
[INFO] |  \- org.osgi:osgi.cmpn:jar:6.0.0:compile
[INFO] +- org.wso2.orbit.xerces:xercesImpl:jar:2.12.0.wso2v1:compile
[INFO] \- org.wso2.config.mapper:config-mapper:jar:1.0.12:compile
[INFO]    +- com.hubspot.jinjava:jinjava:jar:2.4.14:compile
[INFO]    |  +- org.slf4j:slf4j-api:jar:1.7.21:compile
[INFO]    |  +- <b>com.google.guava:guava:jar:27.0-jre:compile (version managed from 22.0)</b>
[INFO]    |  |  +- com.google.guava:failureaccess:jar:1.0:compile
[INFO]    |  |  +- com.google.guava:listenablefuture:jar:9999.0-empty-to-avoid-conflict-with-guava:compile
[INFO]    |  |  +- org.checkerframework:checker-qual:jar:2.5.2:compile
[INFO]    |  |  +- com.google.errorprone:error_prone_annotations:jar:2.2.0:compile
[INFO]    |  |  +- com.google.j2objc:j2objc-annotations:jar:1.1:compile
[INFO]    |  |  \- org.codehaus.mojo:animal-sniffer-annotations:jar:1.17:compile
[INFO]    |  +- org.javassist:javassist:jar:3.22.0-GA:compile
[INFO]    |  +- org.jsoup:jsoup:jar:1.10.3:compile
[INFO]    |  +- com.google.re2j:re2j:jar:1.2:compile
[INFO]    |  +- org.apache.commons:commons-lang3:jar:3.4:compile
[INFO]    |  +- com.google.code.findbugs:annotations:jar:3.0.1:compile
[INFO]    |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.10.3:compile (version managed from 2.7.9.1)
[INFO]    |  |  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.10.3:compile
[INFO]    |  |  \- (com.fasterxml.jackson.core:jackson-core:jar:2.10.3:compile - version managed from 2.7.9; omitted for duplicate)
[INFO]    |  \- com.fasterxml.jackson.core:jackson-core:jar:2.10.3:compile
[INFO]    +- net.consensys.cava:cava-toml:jar:0.5.0:compile
[INFO]    |  +- org.antlr:antlr4-runtime:jar:4.7.1:compile
[INFO]    |  \- com.google.code.findbugs:jsr305:jar:3.0.2:compile
[INFO]    +- com.google.code.gson:gson:jar:2.8.5:compile
[INFO]    +- commons-logging:commons-logging:jar:1.2:compile
[INFO]    +- commons-lang.wso2:commons-lang:jar:2.6.0.wso2v1:compile
[INFO]    \- commons-codec:commons-codec:jar:1.12:compile
</code></pre>

**Comment by Bing-ok** [comment]
@Nirothipan I developed a tool that can automatically detect dependency conflict issues. I can share you a download link later. Could you help me review this issue? Thank you very much.

**Comment by Nirothipan** [comment]
@Bing-ok look at the analysis shared, I could see the possibility of the no such method error. However need to analyze further on the possibility of those in the runtime as we haven't encountered this so far. Thanks.

**Comment by Bing-ok** [comment]
@Nirothipan Thank you very much for your feedback.
Guava always introduces dependency hell due to its bad compatibilty. Our analyzer indeed captures six missing methods (hidden risk) caused by multiple versions of such library. And I provided an another fixing solution to solve the problem.
> Upgrade dependency com.hubspot.jinjava:jinjava from 2.4.14 to 2.4.15, by declaring a direct dependency com.hubspot.jinjava:jinjava:2.4.15 in the pom file of micro-integrator\components\org.wso2.micro.integrator.server. 

Because the newer version com.hubspot.jinjava:jinjava:2.4.15 does not invoke the above missing methods, such upgrading can solve the problem.

I have verified that such upgrading will not introduce other issues.
